gateway/gateway.go

```diff
@@ -787,6 +788,75 @@ func (gateway *HandleT) pendingEventsHandler(w http.ResponseWriter, r *http.Requ
 	w.Write([]byte(fmt.Sprintf("{ \"pending_events\": %d }", totalPendingCount)))
 }
 
+type failedEventsRequestPayload struct {
+	TaskRunID string `json:"task_run_id"`
+}
+
+func (gateway *HandleT) fetchFailedEventsHandler(w http.ResponseWriter, r *http.Request) {
+	gateway.failedEventsHandler(w, r, "fetch")
+}
+
+func (gateway *HandleT) clearFailedEventsHandler(w http.ResponseWriter, r *http.Request) {
+	gateway.failedEventsHandler(w, r, "clear")
+}
+
+func (gateway *HandleT) failedEventsHandler(w http.ResponseWriter, r *http.Request, reqType string) {
+	gateway.logger.LogRequest(r)
+	atomic.AddUint64(&gateway.recvCount, 1)
+	var errorMessage string
+	defer func() {
+		if errorMessage != "" {
+			gateway.logger.Debug(errorMessage)
+			http.Error(w, response.GetStatus(errorMessage), 400)
+		}
+	}()
+
+	payload, _, err := gateway.getPayloadAndWriteKey(w, r, reqType)
+	if err != nil {
+		errorMessage = err.Error()
+		return
+	}
+
+	if !gjson.ValidBytes(payload) {
+		errorMessage = response.GetStatus(response.InvalidJSON)
+		return
+	}
+
+	var reqPayload failedEventsRequestPayload
+	err = json.Unmarshal(payload, &reqPayload)
+	if err != nil {
+		errorMessage = err.Error()
+		return
+	}
+
+	if reqPayload.TaskRunID == "" {
+		errorMessage = "Empty task run id"
+		return
+	}
+
+	if reqType == "fetch" {
+		failedEvents := router.GetFailedEventsManager().FetchFailedRecordIDs(reqPayload.TaskRunID)
+		failedMsgIDsByDestinationID := make(map[string][]string)
+		for _, failedEvent := range failedEvents {
+			if _, ok := failedMsgIDsByDestinationID[failedEvent.DestinationID]; !ok {
+				failedMsgIDsByDestinationID[failedEvent.DestinationID] = []string{}
+			}
+			failedMsgIDsByDestinationID[failedEvent.DestinationID] = append(failedMsgIDsByDestinationID[failedEvent.DestinationID], failedEvent.RecordID)
+		}
+
+		resp, err := json.Marshal(failedMsgIDsByDestinationID)
+		if err != nil {
+			errorMessage = err.Error()
+			return
+		}
+
+		w.Write(resp)
+	} else if reqType == "clear" {
+		router.GetFailedEventsManager().DropFailedRecordIDs(reqPayload.TaskRunID)
+		w.Write([]byte("OK"))
```

Comment:
Change If-ElseIf to switch case
Helper function fo build `failedMsgIDsByDestinationID`